src/middlewared/middlewared/plugins/container/migrate.py

```diff
@@ -1,15 +1,133 @@
+    v4_network = sa.Column(sa.String(120), nullable=True)
+    v6_network = sa.Column(sa.String(120), nullable=True)
 
 
 class ContainerService(Service):
+
+    @private
+    async def migrate_devices(self, job, manifest, container_instance):
+        devices = manifest["devices"]
+        container_name = container_instance["name"]
+        nic_choices = await self.middleware.call("container.device.nic_attach_choices")
+        gpu_choices = await self.middleware.call("container.device.gpu_choices")
+        for device_name, device_data in devices.items():
+            dtype = None
+            try:
+                device_payload = None
+                dtype = device_data.get("type")
+                if dtype == "disk":
+                    src = device_data.get("source", "")
+                    if src.startswith("/mnt") is False:
+                        job.logs_fd.write((
+                            f"Skipping migrating {device_name!r} disk device for {container_name!r} because "
+                            f"source does not start with /mnt/ (is {src!r} instead)\n"
+                        ).encode())
+                        continue
+
+                    device_payload = {
+                        "dtype": "FILESYSTEM",
+                        "source": src,
+                        "target": device_data["path"],
+                    }
+                elif dtype == "nic":
+                    if device_data.get("parent") not in nic_choices:
+                        job.logs_fd.write((
+                            f"Skipping migrating {device_name!r} NIC device for {container_name!r} because "
+                            f"{device_data.get('parent')!r} is not a valid NIC\n"
+                        ).encode())
+                        continue
+
+                    device_payload = {
+                        "dtype": "NIC",
+                        "nic_attach": device_data["parent"],
+                        "type": "VIRTIO",
+                        "trust_guest_rx_filters": False,
+                        "mac": manifest["config"].get(f"volatile.{device_data['parent']}.hwaddr")
+                    }
+                elif dtype == "usb":
+                    if (bus_num := device_data.get("busnum")) and (devnum := device_data.get("devnum")):
+                        device_payload = {
+                            "dtype": "USB",
+                            "device": f"usb_{bus_num}_{devnum}",
+                            "usb": None,
+                        }
+                    elif (vendor_id := device_data.get("vendorid")) and (product_id := device_data.get("productid")):
+                        device_payload = {
+                            "dtype": "USB",
+                            "usb": {"vendor_id": f"0x{vendor_id}", "product_id": f"0x{product_id}"},
+                            "device": None
+                        }
+                    else:
+                        job.logs_fd.write((
```

Comment:
Is this potentially blocking? I think you probably need to either switch this method to sync or do all this fd writing in a thread.